File: src/filecleaner/leftovers.py
"""Two specific kinds of "installation residue" scanning's glob-based rules
can't express, since both need to cross-reference *other* filesystem state
(what's actually installed) rather than matching a fixed path pattern:

1. **Orphaned app-support folders**: subfolders under ``~/Library/{Application
   Support,Caches,Preferences,...}`` whose owning app is no longer in
   ``/Applications``. Folder-naming heuristics have real false-positive
   potential (a Mac App Store sandbox container, a CLI tool with no ``.app``
   bundle) — this is why results are risk ``"high"`` and callers should treat
   it as opt-in, same spirit as the opt-in rules in ``rules.py``.
2. **Installer archive cleanup**: ``.dmg``/``.pkg``/``.zip`` files in
   Downloads whose apparent product already exists — either an installed
   ``.app`` with a matching name, or an already-extracted sibling folder
   next to the archive.

Both are read-only: they return ``Candidate`` objects, exactly like
``scanner.run_scan``, so callers act on them through the normal
``quarantine_candidates``/``purge_entries`` flow — nothing here ever
touches the filesystem.
"""

from __future__ import annotations

import os
import time
from pathlib import Path
from typing import Any

from filecleaner import config as config_mod
from filecleaner import listing, plists, safety
from filecleaner.models import Candidate
from filecleaner.scanner import dir_stats_many

_LIBRARY_SUBDIRS = (
    "Application Support",
    "Caches",
    "Preferences",
    "Logs",
    "Saved Application State",
    "Containers",
    "HTTPStorages",
)
_APPLICATIONS_DIRS = (Path("/Applications"), Path.home() / "Applications")
_INSTALLER_EXTENSIONS = (".dmg", ".pkg", ".zip")
_APP_LEFTOVERS_CATEGORY = "App leftovers (opt-in)"
_INSTALLER_CLEANUP_CATEGORY = "Installer cleanup (opt-in)"


def installed_apps() -> tuple[set[str], set[str]]:
    """(bundle_ids, display_names) for every ``.app`` under ``/Applications``
    and ``~/Applications`` — the reference set both detectors match against."""
    bundle_ids: set[str] = set()
    names: set[str] = set()
    for apps_dir in _APPLICATIONS_DIRS:
        try:
            entries = list(apps_dir.iterdir())
        except OSError:
            continue
        for entry in entries:
            if entry.suffix != ".app":
                continue
            names.add(entry.stem)
            bundle_id = plists.load_dict(entry / "Contents" / "Info.plist").get("CFBundleIdentifier")
            if isinstance(bundle_id, str):
                bundle_ids.add(bundle_id)
    return bundle_ids, names


def _looks_installed(name: str, bundle_ids: set[str], names: set[str]) -> bool:
    if name in bundle_ids or name in names:
        return True
    # Many Library subfolders are named after the bundle id with extra
    # suffixes, e.g. "com.example.app.savedState" — match on prefix too.
    return any(name.startswith(bid) for bid in bundle_ids)


def find_app_leftovers(
    config: dict[str, Any], *, home: Path | None = None, min_age_days: int = 30
) -> list[Candidate]:
    """Subfolders under ``~/Library/{Application Support,Caches,...}`` whose
    name matches no currently-installed app. Opt-in, high risk."""
    home = home or Path.home()
    bundle_ids, names = installed_apps()
    extra_protected = config_mod.extra_protected_paths(config) + config_mod.data_paths_to_protect(config)
    now = time.time()

    orphans: list[tuple[Path, os.stat_result, bool]] = []
    for subdir_name in _LIBRARY_SUBDIRS:
        base = home / "Library" / subdir_name
        try:
            # Containers is where macOS interrupts listings (see ``listing``).
            entries = [base / entry.name for entry in listing.list_dir(str(base))]
        except OSError:
            continue
        for entry in entries:
            if entry.is_symlink() or safety.is_protected(entry, extra_protected=extra_protected):
                continue
            if _looks_installed(entry.name, bundle_ids, names):
                continue
            try:
                st = entry.lstat()
            except OSError:
                continue
            orphans.append((entry, st, entry.is_dir()))

    # Sizing the folders is nearly all of the work (a thousand of them, some
    # huge), so it is done in one batch that the native helper can spread out.
    sized = iter(dir_stats_many([entry for entry, _st, is_dir in orphans if is_dir]))
    candidates: list[Candidate] = []
    for entry, st, is_dir in orphans:
        if is_dir:
            size, newest = next(sized)
            mtime = max(st.st_mtime, newest)
        else:
            size, mtime = st.st_size, st.st_mtime
        if (now - mtime) / 86400 < min_age_days:
            continue
        candidates.append(
            Candidate(
                path=entry,
                size_bytes=size,
                is_dir=is_dir,
                mtime=mtime,
                rule_id="app_leftovers",
                category=_APP_LEFTOVERS_CATEGORY,
                risk="high",
            )
        )
    return candidates

def find_installer_cleanup(config: dict[str, Any], *, downloads: Path | None = None) -> list[Candidate]:
    """``.dmg``/``.pkg``/``.zip`` files in Downloads whose apparent product
    already exists (a matching-name installed ``.app``, or an
    already-extracted sibling folder). Opt-in, medium risk."""
    downloads = downloads or (Path.home() / "Downloads")
    bundle_ids, names = installed_apps()
    extra_protected = config_mod.extra_protected_paths(config) + config_mod.data_paths_to_protect(config)
    candidates: list[Candidate] = []

    try:
        entries = list(downloads.iterdir())
    except OSError:
        return []

    for entry in entries:
        if entry.is_symlink() or not entry.is_file():
            continue
        if entry.suffix.lower() not in _INSTALLER_EXTENSIONS:
            continue
        if safety.is_protected(entry, extra_protected=extra_protected):
            continue
        stem = entry.stem
        already_installed = stem in names or any(bid.rsplit(".", 1)[-1].lower() == stem.lower() for bid in bundle_ids)
        already_extracted = (downloads / stem).is_dir()
        if not (already_installed or already_extracted):
            continue
        try:
            st = entry.stat()
        except OSError:
            continue
        candidates.append(
            Candidate(
                path=entry,
                size_bytes=st.st_size,
                is_dir=False,
                mtime=st.st_mtime,
                rule_id="installer_cleanup",
                category=_INSTALLER_CLEANUP_CATEGORY,
                risk="medium",
            )
        )
    return candidates

File: native/fclean-walk/src/leftovers.rs
//! `leftovers.find_app_leftovers`, `leftovers.find_installer_cleanup` and what
//! `fclean leftovers --json` prints. Finding only: `--apply` is not ported.
//!
//! Both detectors cross-reference what is installed, which no glob can
//! express. They are heuristics, and are ported as they are — including the
//! order of what they find, which is the order the folders list in.

use std::collections::HashSet;
use std::fs;
use std::time::{SystemTime, UNIX_EPOCH};

use rayon::prelude::*;
use serde::Deserialize;

use crate::pyjson::{dumps, J};
use crate::pypath::{normalise, splitext};
use crate::report::{load, Loaded, Setup};
use crate::{config, dir_stats, fail, mtime_nanos, read_dir, safety, st_mtime, start_pool, Deny};

const LIBRARY_SUBDIRS: [&str; 7] =
    ["Application Support", "Caches", "Preferences", "Logs", "Saved Application State", "Containers", "HTTPStorages"];
const INSTALLER_EXTENSIONS: [&str; 3] = [".dmg", ".pkg", ".zip"];
const MIN_AGE_DAYS: f64 = 30.0;

#[derive(Deserialize)]
struct Request {
    #[serde(flatten)]
    setup: Setup,
    /// `fclean leftovers --kind apps|installers|all`.
    #[serde(default = "default_kind")]
    kind: String,
    /// Python looks these up once, when `leftovers` is imported.
    applications_dirs: Option<Vec<String>>,
    now: Option<f64>,
}

fn default_kind() -> String {
    "all".to_owned()
}

pub struct Candidate {
    pub path: String,
    pub size: u64,
    pub is_dir: bool,
    pub mtime: f64,
    pub rule_id: &'static str,
    pub category: &'static str,
    pub risk: &'static str,
}

impl Candidate {
    /// `Candidate.to_dict`.
    pub fn to_json(&self) -> J {
        J::dict([
            ("path", J::str(&self.path)),
            ("size_bytes", J::UInt(self.size)),
            ("is_dir", J::Bool(self.is_dir)),
            ("mtime", J::Float(self.mtime)),
            ("rule_id", J::str(self.rule_id)),
            ("category", J::str(self.category)),
            ("risk", J::str(self.risk)),
        ])
    }
}

pub struct Installed {
    bundle_ids: HashSet<String>,
    names: HashSet<String>,
}

/// The names in `dir`, in the order it lists them. A name that is not
/// Unicode cannot be one Python would have printed, and is left out.
fn names_in(dir: &str) -> Option<Vec<String>> {
    Some(read_dir(dir).ok()?.into_iter().filter_map(|entry| entry.file_name().into_string().ok()).collect())
}

/// `installed_apps`: every `.app` in the applications folders, by bundle
/// identifier and by name. An `Info.plist` that cannot be read names nothing.
pub fn installed_apps(applications_dirs: &[String]) -> Installed {
    let mut installed = Installed { bundle_ids: HashSet::new(), names: HashSet::new() };
    for dir in applications_dirs {
        for name in names_in(dir).unwrap_or_default() {
            let (stem, suffix) = splitext(&name);
            if suffix != ".app" {
                continue;
            }
            installed.names.insert(stem.to_owned());
            let Ok(info) = plist::Value::from_file(format!("{dir}/{name}/Contents/Info.plist")) else { continue };
            let bundle_id = info.as_dictionary().and_then(|info| info.get("CFBundleIdentifier")).and_then(|id| id.as_string());
            if let Some(bundle_id) = bundle_id {
                installed.bundle_ids.insert(bundle_id.to_owned());
            }
        }
    }
    installed
}

impl Installed {
    /// `_looks_installed`. Many Library folders are named after the bundle
    /// identifier with something added ("com.example.app.savedState").
    fn owns(&self, name: &str) -> bool {
        self.bundle_ids.contains(name) || self.names.contains(name) || self.bundle_ids.iter().any(|id| name.starts_with(id.as_str()))
    }
}

/// Folders and files under `~/Library/{Application Support,Caches,...}` that
/// no installed app accounts for, untouched for a month. Opt-in, high risk.
pub fn find_app_leftovers(home: &str, installed: &Installed, index: &Deny, cwd: &str, now: f64) -> Vec<Candidate> {
    let mut orphans: Vec<(String, fs::Metadata)> = Vec::new();
    for subdir in LIBRARY_SUBDIRS {
        let base = format!("{home}/Library/{subdir}");
        for name in names_in(&base).unwrap_or_default() {
            let path = format!("{base}/{name}");
            let meta = fs::symlink_metadata(&path);
            if meta.as_ref().is_ok_and(|meta| meta.file_type().is_symlink()) || safety::is_protected(&path, index, cwd) {
                continue;
            }
            if installed.owns(&name) {
                continue;
            }
            if let Ok(meta) = meta {
                orphans.push((path, meta));
            }
        }
    }

    // `scanner.dir_stats_many`: sized together, across every core. Python
    // asks the helper only when there is more than one folder, and sizes a
    // lone one itself — reading the newest time its own way (`st_mtime`).
    let folders: Vec<&(String, fs::Metadata)> = orphans.iter().filter(|(_, meta)| meta.is_dir()).collect();
    let by_helper = folders.len() > 1;
    let mut sized = folders.par_iter().map(|(path, meta)| dir_stats(path, mtime_nanos(meta))).collect::<Vec<_>>().into_iter();

    let mut candidates = Vec::new();
    for (path, meta) in &orphans {
        let own = st_mtime(mtime_nanos(meta));
        let (size, mtime) = match meta.is_dir().then(|| sized.next()).flatten() {
            Some((size, newest)) => (size, own.max(if by_helper { newest as f64 / 1e9 } else { st_mtime(newest) })),
            None => (meta.len(), own),
        };
        if (now - mtime) / 86400.0 < MIN_AGE_DAYS {
            continue;
        }
        candidates.push(Candidate {
            path: path.clone(),
            size,
            is_dir: meta.is_dir(),
            mtime,
            rule_id: "app_leftovers",
            category: "App leftovers (opt-in)",
            risk: "high",
        });
    }
    candidates
}

/// `.dmg`/`.pkg`/`.zip` files in Downloads whose product is already there:
/// an installed app of that name, or a folder of that name next to the
/// archive. Opt-in, medium risk.
pub fn find_installer_cleanup(downloads: &str, installed: &Installed, index: &Deny, cwd: &str) -> Vec<Candidate> {
    let mut candidates = Vec::new();
    for name in names_in(downloads).unwrap_or_default() {
        let path = format!("{downloads}/{name}");
        if !fs::symlink_metadata(&path).is_ok_and(|meta| meta.is_file()) {
            continue; // a symlink, or not a file
        }
        let (stem, suffix) = splitext(&name);
        if !INSTALLER_EXTENSIONS.contains(&suffix.to_lowercase().as_str()) || safety::is_protected(&path, index, cwd) {
            continue;
        }
        let wanted = stem.to_lowercase();
        let already_installed = installed.names.contains(stem)
            || installed.bundle_ids.iter().any(|id| id.rsplit('.').next().is_some_and(|last| last.to_lowercase() == wanted));
        let already_extracted = fs::metadata(format!("{downloads}/{stem}")).is_ok_and(|meta| meta.is_dir());
        if !(already_installed || already_extracted) {
            continue;
        }
        let Ok(meta) = fs::metadata(&path) else { continue };
        candidates.push(Candidate {
            path,
            size: meta.len(),
            is_dir: false,
            mtime: st_mtime(mtime_nanos(&meta)),
            rule_id: "installer_cleanup",
            category: "Installer cleanup (opt-in)",
            risk: "medium",
        });
    }
    candidates
}

pub fn main(input: &str) {
    let request: Request =
        serde_json::from_str(input).unwrap_or_else(|err| fail(format!("bad leftovers-json request: {err}")));
    if !["apps", "installers", "all"].contains(&request.kind.as_str()) {
        fail(format!("--kind must be 'apps', 'installers', or 'all', got {}", crate::rules::repr(&request.kind)));
    }
    let Loaded { env, paths, config, .. } = load(&request.setup);
    let home = normalise(&env.home);
    let extra_protected = config::protected_paths(&config, &paths, &env.home).unwrap_or_else(|err| fail(err));
    let index = safety::build_index(&env, &extra_protected);
    let applications_dirs =
        request.applications_dirs.unwrap_or_else(|| vec!["/Applications".to_owned(), format!("{home}/Applications")]);
    let now = request
        .now
        .unwrap_or_else(|| SystemTime::now().duration_since(UNIX_EPOCH).map(|since| since.as_secs_f64()).unwrap_or(0.0));
    start_pool(request.setup.threads);

    let installed = installed_apps(&applications_dirs);
    let mut candidates = Vec::new();
    if request.kind != "installers" {
        candidates.extend(find_app_leftovers(&home, &installed, &index, &env.cwd, now));
    }
    if request.kind != "apps" {
        candidates.extend(find_installer_cleanup(&format!("{home}/Downloads"), &installed, &index, &env.cwd));
    }
    let report = J::dict([
        ("candidates", J::List(candidates.iter().map(Candidate::to_json).collect())),
        ("total_size_bytes", J::UInt(candidates.iter().map(|candidate| candidate.size).sum())),
    ]);
    println!("{}", dumps(&report));
}
